omniscidb/Tests/ResultSetArrowConversion.cpp

```diff
@@ -385,6 +387,231 @@ TEST(ArrowTable, EmptySelection) {
+}
+
+TEST(ArrowTable, ColumnarLimitedJoin) {
+  bool prev_enable_columnar_output = config().rs.enable_columnar_output;
+  bool prev_enable_lazy_fetch = config().rs.enable_lazy_fetch;
+
+  ScopeGuard reset = [prev_enable_columnar_output, prev_enable_lazy_fetch] {
+    config().rs.enable_columnar_output = prev_enable_columnar_output;
+    config().rs.enable_lazy_fetch = prev_enable_lazy_fetch;
+  };
+
+  config().rs.enable_columnar_output = true;
+  config().rs.enable_lazy_fetch = false;
+  auto res = runSqlQuery(
+      "SELECT * FROM test_chunked INNER JOIN join_table ON test_chunked.i=join_table.i "
+      "offset 5 limit 2;",
+      ExecutorDeviceType::CPU,
+      true);
+
+  ASSERT_EQ(res.getRows()->rowCount(), (int64_t)1);
+
+  auto table = getArrowTable(res);
+  ASSERT_NE(table, nullptr);
+  ASSERT_EQ(table->num_columns(), 7);
+  ASSERT_EQ(table->num_rows(), (int64_t)1);
+
+  std::array<int64_t, 1> res_arr_1{1};
+  compare_columns(res_arr_1, table->column(1));
+
+  std::array<int64_t, 1> res_arr_2{6};
+  compare_columns(res_arr_2, table->column(2));
+
+  std::array<double, 1> res_arr_3{60.6};
+  compare_columns(res_arr_3, table->column(3));
+
+  std::array<int64_t, 1> res_arr_4{1};
+  compare_columns(res_arr_4, table->column(4));
+
+  // compare_columns fails on bool
+  // std::array<bool, 1> res_arr_5{true};
+  // compare_columns(res_arr_5, table->column(5));
+
+  std::array<int64_t, 1> res_arr_6{200};
+  compare_columns(res_arr_6, table->column(6));
+}
+
+TEST(ArrowTable, ColumnarEmptyLimitedJoin) {
+  bool prev_enable_columnar_output = config().rs.enable_columnar_output;
+  bool prev_enable_lazy_fetch = config().rs.enable_lazy_fetch;
+
+  ScopeGuard reset = [prev_enable_columnar_output, prev_enable_lazy_fetch] {
+    config().rs.enable_columnar_output = prev_enable_columnar_output;
+    config().rs.enable_lazy_fetch = prev_enable_lazy_fetch;
+  };
+
+  config().rs.enable_columnar_output = true;
+  config().rs.enable_lazy_fetch = false;
+  auto res = runSqlQuery(
+      "SELECT * FROM test_chunked offset 8 limit 2;", ExecutorDeviceType::CPU, true);
+
+  ASSERT_EQ(res.getRows()->rowCount(), (int64_t)0);
+
+  auto table = getArrowTable(res);
+  ASSERT_NE(table, nullptr);
+  ASSERT_EQ(table->num_columns(), 4);
+  ASSERT_EQ(table->num_rows(), (int64_t)0);
+}
+
+TEST(ArrowTable, ColumnarMultiStoragedJoin) {
+  bool prev_enable_columnar_output = config().rs.enable_columnar_output;
+  bool prev_enable_lazy_fetch = config().rs.enable_lazy_fetch;
+
+  ScopeGuard reset = [prev_enable_columnar_output, prev_enable_lazy_fetch] {
+    config().rs.enable_columnar_output = prev_enable_columnar_output;
+    config().rs.enable_lazy_fetch = prev_enable_lazy_fetch;
+  };
+
+  config().rs.enable_columnar_output = true;
+  config().rs.enable_lazy_fetch = false;
+  auto res = runSqlQuery(
+      "SELECT * FROM test_chunked offset 3 limit 7;", ExecutorDeviceType::CPU, true);
+
+  ASSERT_EQ(res.getRows()->rowCount(), (int64_t)3);
```

Comment:
I definitely see problems in the current columnar conversion implementation. That makes me think these tests don't actually give you the required result set layouts. Please add asserts for `isChunkedZeroCopyColumnarConversionPossible()`, `getLimit()`, and `getOffset()` to make sure you have what you expect.

Reply:
donedone